src/pool.rs

```diff
@@ -211,12 +242,31 @@ impl ConnectionPool {
                             replica_number += 1;
                         }
 
+                        // We assume every server in the pool share user/passwords
```

Comment:
They don't :)

Reply:
Uhmm, I think we are mixing three things here

**First, the whole auth_query per pool:**
I understand (and agree) what you say about "one auth_query per pool". I can re-implement the think to be like that, I think the implementation should be done using the config at the general scope and inherit that (so we do not have to be copying the thing in every single pool), the same way it is done for `pool_size`.

Given that MD5 hashes are currently being stored at the pool level, implementing this should be straightforward, just dealing with the `config.rs` part.

**Second, Pool::auth_hash_has_changed:**
What this do is it checks whether the password has changed in the server and then updates the info at the pool level. This is the implementation of:

> We could try to do what we do for our config validation: fetch all information at boot and store it in a `HashMap` or similar. When a user attempts to login, we compare their md5 hash with what we fetched from `pg_shadow` on boot. It won't be real time, e.g. when a new user is added, this mapping has to be refetched, but that can be implemented as part of `RELOAD`.
from [here](https://github.com/levkk/pgcat/issues/253#issuecomment-1356854114).

Implementing password change also implements the adding new user 'use-case'. This is tested and works.

**Third: We assume every server in the pool share user/passwords**
What is assumed here, is that every server in the connection pool (a connection pool being 1 user, 1 database n servers) shares the same user/password, and this is an assumption that is already being made in current implementation (with cleartext), also this assumption is logical, because you are in the end connecting to the pool using just one "user/database/password". The only thing that I think we can do is to tell that in logs so the administrator can spot the issue in the server and update its config, which is already done [here](https://github.com/OneSignal/pgcat/blob/auth-query/src/pool.rs#L253)